internal/bootstrap/bootstrap.go

```diff
@@ -1,22 +1,143 @@
+	"github.com/gardener/etcd-wrapper/internal/util"
+	"go.etcd.io/etcd/server/v3/embed"
+	"go.uber.org/zap"
+)
+
+const (
+	defaultBackupRestoreMaxRetries = 5
+	defaultBackOffBetweenRetries   = 1 * time.Second
 )
 
-// InitializeEtcd triggers the initialization of etcd DB.
-func InitializeEtcd(ctx context.Context) error {
-	//create new BackupRestoreClient
-	brClient := brclient.NewClient("")
-	for {
+// EtcdInitializer is an interface for methods to be used to initialize etcd
+type EtcdInitializer interface {
+	Run(context.Context) (*embed.Config, error)
+}
+
+type initializer struct {
+	brClient brclient.BackupRestoreClient
+	logger   *zap.Logger
+}
+
+// NewEtcdInitializer creates and returns an EtcdInitializer object
+func NewEtcdInitializer(brConfig *types.BackupRestoreConfig, logger *zap.Logger) (EtcdInitializer, error) {
+	// Validate backup-restore configuration
+	if err := brConfig.Validate(); err != nil {
+		return nil, err
+	}
+
+	//create backup-restore client
+	brClient, err := brclient.NewDefaultClient(*brConfig, brclient.DefaultEtcdConfigFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &initializer{
+		brClient: brClient,
+		logger:   logger,
+	}, nil
+}
+
+// Run initializes the etcd and gets the etcd configuration
+func (i *initializer) Run(ctx context.Context) (*embed.Config, error) {
+	var (
+		err        error
+		initStatus brclient.InitStatus
+	)
+	for initStatus != brclient.Successful {
+		if initStatus, err = i.brClient.GetInitializationStatus(ctx); err != nil {
+			i.logger.Error("error while fetching initialization status", zap.Error(err))
+		}
+		i.logger.Info("Fetched initialization status", zap.String("Status", initStatus.String()))
+		if initStatus == brclient.New {
+			validationMode := determineValidationMode(types.DefaultExitCodeFilePath, i.logger)
+			i.logger.Info("Fetched initialization status is `New`. Triggering etcd initialization with validation mode", zap.Any("mode", validationMode))
+			if err = i.brClient.TriggerInitialization(ctx, validationMode); err != nil {
+				i.logger.Error("error while triggering initialization to backup-restore", zap.Error(err))
+			}
+		}
 		select {
 		case <-ctx.Done():
-			return ctx.Err()
-		default:
+			return nil, ctx.Err()
+		case <-time.After(defaultBackOffBetweenRetries):
+		}
+	}
+	i.logger.Info("Etcd initialization succeeded")
+	return i.tryGetEtcdConfig(ctx, defaultBackupRestoreMaxRetries, defaultBackOffBetweenRetries)
+}
+
+// CaptureExitCode captures the exit signal into a file `exit_code`
+func CaptureExitCode(signal os.Signal, exitCodeFilePath string) error {
+	if signal == nil {
+		return nil
+	}
+	interruptSignal := []byte(signal.String())
+	return os.WriteFile(exitCodeFilePath, interruptSignal, 0644)
+}
+
+// CleanupExitCode removes the `exit_code` file
+func CleanupExitCode(exitCodeFilePath string) error {
+	err := os.Remove(exitCodeFilePath)
+	if errors.Is(err, os.ErrNotExist) {
+		//log file does not exist
```

Comment:
Please remove this comment or you can mark as TODO if you want to handle this in future

Reply:
this comment is still not addressed.

Reply:
removed it.